DataExporter: Enhance DataExporter to support other components #4414

ravihariharan opened the pull request:
DataExporter: Enhance DataExporter to support other components #4414 

Closed Old PR #4416

ravihariharan committed:
DataExporter: Enhance DataExporter to support other components #4414

ravihariharan committed:
DataExporter: Enhance DataExporter to support other components #4414

Review By Rapster #4416
Fix License

Rapster commented:
Please don't change the format

ravihariharan commented:
@Rapster  which format ? i cherry picked the commit and merge the code in a new branch. I also had to fix the License issue.

Rapster commented:
See `Exporter#export()` parameters were previously well formatted

ravihariharan committed:
DataExporter: Enhance DataExporter to support other components #4414

Review Rapster primefaces#4949

Rapster commented:
I guess exporters should be renamed then, for instance `ExcelExporter` -> `ExcelDataTableExporter`?

ravihariharan commented:
> I guess exporters should be renamed then, for instance `ExcelExporter` -> `ExcelDataTableExporter`?

The current changes are backward compatible. 

Since ExcelExporter and other were not final class i did not rename them.  
An example of breaking change would be extending ExcelExporter  and using a customExporter.
https://github.com/primefaces/primefaces/issues/4397

Rapster commented:
Sorry I don't understand your point

ravihariharan commented:
@Rapster  I made only the minimum changes that were required to support extending customExporter to be used with other components.  

We can rename all the class or 
We can move the DataTable specific classes to subpackage of .export.datatable
![image](https://user-images.githubusercontent.com/15189434/60354109-83b16d80-9999-11e9-8576-379fbaac21af.png)

Lets see what @tandraschko / @melloware / @mertsincan   has to say about rename or moving these classes.

Rapster commented:
+1 for creating the appropriate package, but I don't think all of them need to move in that new package. Few of these classes can be reused no matter what the component is

ravihariharan commented:
> +1 for creating the appropriate package, but I don't think all of them need to move in that new package

All the class in the sub-package are specific to DataTable. (moved *Options class back)
![image](https://user-images.githubusercontent.com/15189434/60355696-0c7dd880-999d-11e9-9b37-2d6e4df18c30.png)

ExporterFactory can be renamed to DataTableExporterFactory.

Rapster commented:
Why callbacks and factory should be different from the rest? (This is a trick question btw :yum: )

ravihariharan commented:
> Why callbacks and factory should be different from the rest?



![image](https://user-images.githubusercontent.com/15189434/60356228-76e34880-999e-11e9-8573-3257cf3ddb49.png)


We will have to remove <T extends UIComponent> from Exporter.
Will have to change the methods to have UIComponent  and cast when required.

Let me make few changes for factory  and in another branch and confirm.

Rapster commented:
Well, nothing prevents you to create a `PDFExporter` interface, your "PDFDataTableExporter" implements it and that should solve your problem no?

melloware commented:
I like all these changes.  And it will make future flexibility better.

tandraschko commented:
+1 for the sub-package
and we should also name e.g. XMLExporter -> XMLDataTableExporter

ravihariharan commented:
There are few discussion in https://github.com/primefaces/primefaces/issues/4414

@Rapster  approach was
Keep in mind, that my objective is to be able to reuse these exporters by doing so:
<p:dataExporter type="xls" target="tbl" fileName="cars" pageOnly="true"/>


I was going in the direction of using customExporter for other components. 

e.g. if we want to export a quickbooks format then we could do something like
<p:dataExporter type="quickbooks" customExporter="#{quickbooksExporter}" />
type="quickbooks" is only added for readability


i did not plan for a generic export for treetable , but allow developer to use the customExporter to implement the needed feature.

ravihariharan committed:
Merge pull request #1 from primefaces/master

Merge for ravihariharan/primefaces

Commit:
Merge remote-tracking branch 'origin/master' into Issue_4414

Conflicts:
	src/main/java/org/primefaces/component/export/Exporter.java

Commit:
Fixed #4970

re-applied fix due to conflict

ravihariharan committed:
Enhance DataExporter to support Other Data Components #4414

Move DataTable to subpackage

ravihariharan committed:
Enhance DataExporter to support Other Data Components #4414

Rename to specific implementation

ravihariharan committed:
Enhance DataExporter to support Other Data Components #4414

Review fix : Rapster and tandraschko

ravihariharan commented:
![image](https://user-images.githubusercontent.com/15189434/61880977-df511700-aec3-11e9-8a47-a3b480cea06b.png)


CSVExporter ,  ExcelExporter , PDFExporter are now interface that are referred with corresponding VisitCallback

ravihariharan committed:
Enhance DataExporter to support Other Data Components #4414

Add ExporterBase

ravihariharan committed:
Enhance DataExporter to support Other Data Components #4414

No need of DataTableExporterBase to implement Exporter interface.
Every exported will implemented specific type.

ravihariharan commented:
@tandraschko / @melloware  / @Rapster i have made all the changes as discussed above. Please review and let me know if we can improve something.

1 task for future would be to create a Exporter factory.

melloware commented:
This looks great to me.

Rapster commented:
Since a package `datatable` has been created, I would have probably keep the original name. It's a lot of datatable-something in the datatable package if you know what I mean. Otherwise looks good

ravihariharan commented:
This is what i would like to do @Rapster 
1. Move the datatable specific to org.primefaces.component.**datatable.export** package.
2. Rename the files back

Only reference in org.primefaces.component.export.DataExporter will be to the DataTableExporterFactory

![image](https://user-images.githubusercontent.com/15189434/61920462-b6fb0400-af27-11e9-8fb7-9dcd704fb549.png)


i have the changes in another branch and will merge on confirmation :)

Rapster commented:
I like your second suggestion in terms of packages. Indeed we can't have CSVExporter twice, and we can't use fully qualified class name... Let's keep `DataTable` as a prefix

Rapster commented on src/main/java/org/primefaces/component/export/datatable/DataTableExporterBase.java:
```diff
@@ -0,0 +1,333 @@
+import org.primefaces.component.overlaypanel.OverlayPanel;
+import org.primefaces.util.ComponentUtils;
+
+public abstract class DataTableExporterBase {
```
1. Make this implements `Exporter<DataTable>` (and remove this from its children)
2. Rename into `DataTableExporter` (xxxBase is for definition of UIComponent)`

ravihariharan replied:
Rename into DataTableExporter (xxxBase is for definition of UIComponent) - done

ravihariharan replied:
> Make this implements Exporter<DataTable> (and remove this from its children)

even if we do it , we will need the the children to implement specific interface.
e.g.

`public class DataTablePDFExporter extends DataTableExporter implements PDFExporter<DataTable>`

PDFExporter methods are reference from PDFExportVisitCallback

Rapster replied:
I know but still, DataTableExporter is an exporter (abstract but still an exporter :wink:)

Rapster commented on src/main/java/org/primefaces/component/export/datatable/DataTableCSVExporter.java:
```diff
@@ -0,0 +1,251 @@
+                }
+
+                if (textValue != null) {
+                    this.addColumnValue(builder, textValue);
```
Remove unnecessary `this` references

ravihariharan committed:
Enhance DataExporter to support Other Data Components #4414

Move the datatable specific to org.primefaces.component.datatable.export
package

ravihariharan committed:
Enhance DataExporter to support Other Data Components #4414

Remove unnecessary this references. Review Rapster

ravihariharan committed:
Enhance DataExporter to support Other Data Components #4414

Rename into DataTableExporter (xxxBase is for definition of
UIComponent). Review Rapster

ravihariharan committed:
Enhance DataExporter to support Other Data Components #4414

Make DataTableExporter implements Exporter<DataTable> . Review by
Rapster

ravihariharan commented:
@Rapster / @tandraschko  PR ready for merge.

Rapster commented:
I'm cool with it, do you think by any chance you could provide an example in the showcase?

tandraschko commented:
@ravihariharan please also update the documentation (with samples how to use it)

ravihariharan commented:
Migration Guide updated

ravihariharan commented:
https://github.com/primefaces/showcase-facelift/pull/54

I reused a existing image to use for export for text.

![image](https://user-images.githubusercontent.com/15189434/62225949-5f292680-b387-11e9-9a12-7b80ff990a1a.png)
